File: java11/src/main/java/my/rinat/kata/collections/donut/donutkata10/DonutType.java
package my.rinat.kata.collections.donut.donutkata10;

import org.eclipse.collections.impl.utility.ArrayIterate;

public enum DonutType {
    BOSTON_CREAM("BC"),
    GLAZED("G"),
    OLD_FASHIONED("OF"),
    CHOCOLATE_GLAZED("CG"),
    VANILLA_FROSTED("VF"),
    PUMPKIN("P"),
    BLUEBERRY("B"),
    JELLY("J"),
    BAVARIAN_CREAM("BA");

    private String abbreviation;

    DonutType(String abbreviation) {
        this.abbreviation = abbreviation;
    }

    private boolean abbreviationEquals(String abbreviation) {
        return abbreviation.equals(this.abbreviation);
    }

    public static DonutType forAbbreviation(String abbreviation) {
        return ArrayIterate.detectWith(DonutType.values(), DonutType::abbreviationEquals, abbreviation);
    }
}

File: kotlin/kotlin-eclipse-collections-katas/src/main/kotlin/my/rinat/kata/donutkatakotlin/DonutType.kt
@file:Suppress("RedundantVisibilityModifier", "RemoveRedundantQualifierName")

package my.rinat.kata.donutkatakotlin

import org.eclipse.collections.impl.utility.ArrayIterate

enum class DonutType private constructor(private val abbreviation: String) {
    BOSTON_CREAM("BC"),
    GLAZED("G"),
    OLD_FASHIONED("OF"),
    CHOCOLATE_GLAZED("CG"),
    VANILLA_FROSTED("VF"),
    PUMPKIN("P"),
    BLUEBERRY("B"),
    JELLY("J"),
    BAVARIAN_CREAM("BA");

    private fun abbreviationEquals(abbreviation: String): Boolean {
        return abbreviation == this.abbreviation
    }

    companion object {
        fun forAbbreviation(abbreviation: String): DonutType {
            return ArrayIterate.detectWith(
                    DonutType.values(),
                    { obj, abbr -> obj.abbreviationEquals(abbr) },
                    abbreviation)
        }
    }
}
